rfcs/Struct.md

```diff
@@ -0,0 +1,464 @@
+
+### Struct union
+
+A struct union is an abstract type representing that the value may be one of
+many possible `struct`s. Rather than requiring a new type, we can define a
+struct union as a union that only contains structs:
+
+```graphql
+union Paragraph =
+    TextParagraph
+  | PullquoteParagraph
+  | BlockquoteParagraph
+  | TweetParagraph
+  | GalleryParagraph
+```
+
+(A `union` must only consist of object types, or only of structs, never a mix of
+the two.)
+
+Champion's note: We could use a different keyword, such as `structUnion`, to
+enforce this, but I'm not convinced this is necessary.)
+
+Champion's note: I'm very open to dropping struct union and using a oneOf
+approach instead, or adding a oneOf approach in addition. I've gone with struct
+union for now for greater similarity with the existing GraphQL types.
+
+A struct union is valid in both input and output.
+
+On input of a struct union, `__typename` is required in order to determine the
+type of the `struct` supplied.
+
+### Example schema
+
+Here's an example schema where a user can customise their biography by composing
+together paragraphs of different types. You could imagine that these paragraphs
+were managed by an editor such as ProseMirror.
+
+```graphql
+type Query {
+  user(id: ID!): User
+}
+
+type Mutation {
+  setUserBio(userId: ID!, bio: Biography!): User
+}
+
+type User {
+  id: ID!
+  username: String!
+  bio: Biography!
+}
+
+struct Biography {
+  title: String!
+  socials: BiographySocials
+  paragraphs: [Paragraph!]!
+}
+
+struct BiographySocials {
+  github: String
+  twitter: String
+  linkedIn: String
+  facebook: String
+}
+
+union Paragraph =
+  | TextParagraph
+  | PullquoteParagraph
+  | BlockquoteParagraph
+  | TweetParagraph
+  | GalleryParagraph
+
+struct TextParagraph {
+  text: String!
+}
+struct PullquoteParagraph {
+  pullquote: String!
+  source: String
+}
+struct BlockquoteParagraph {
+  paragraphs: [Paragraph!]!
+  source: String
+}
+struct TweetParagraph {
+  message: String!
+  url: String!
+}
+struct GalleryParagraph {
+  images: [Image!]!
+}
+
+struct Image {
+  url: String!
+  caption: String
+}
+```
+
+### Selection sets
+
+Controversially, structs could be the first type in GraphQL that can act as both
+a leaf _and_ a non-leaf type - i.e. a selection set over them is optional. If
+you do not provide a selection set then the entire value is returned. If you do
+provide a selection set then those fields will be selected and returned.
+
+Struct selection sets are similar to, but not the same as, regular selection
+sets. In particular:
+
+- struct fields are not `FIELD`s, i.e. they cannot have `FIELD` directives
+  attached (suggest we call them `STRUCT_FIELD`)
+- aliases are not allowed, otherwise the simple field merging (see below) does
+  not work (also there's no need for aliasing because struct fields are pure
+  data and do not accept arguments, so they cannot produce multiple values for
+  the same field)
+
```

Comment:
that does not seem convincing, to have struct selection field different from regular fields.
No field directives - why not? something like @include? As for aliases, you argue they are not so useful - so what? Forbidding dirs and aliases brings more problems than benefits - users will keep forgetting that cannot use them with structs. Better treat struct fields as regular fields, just with some type restrictions